### PR TITLE
refactor(compiler): update the compiler for algorithm w

### PR DESCRIPTION
**Pull request by jsternberg**
### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: compiler/compiler.go [hunk 1]
```diff
@@ -379,16 +447,21 @@ func compile(n semantic.Node, subst map[uint64]semantic.MonoType, scope Scope) (
...
 		}
-		lt := l.Type()
+		lt := l.Type().Nature()
 		r, err := compile(n.Right, subst, scope)
 		if err != nil {
 			return nil, err
 		}
-		rt := r.Type()
+		rt := r.Type().Nature()
+		if lt == semantic.Invalid {
```
Comment (jlapacik) [inline compiler/compiler.go]: This means the left hand side is null?

File: compiler/compiler.go [hunk 2]
```diff
@@ -91,27 +64,145 @@ func Compile(scope Scope, f *semantic.FunctionExpression, in semantic.MonoType)
...
+		// We can do this by calling substituteTypes with the same
+		// input parameter and the substituted monotype since
+		// substituteTypes will verify the types.
+		if t, ok := subst[vn]; ok {
+			return substituteTypes(subst, t, in)
+		}
+
+		// If the input type is not invalid, mark it down
+		// as the real type.
+		if in.Nature() != semantic.Invalid {
```
[the remark this answers is not in the document]
Comment (jsternberg) [reply compiler/compiler.go]: Yes. That was my intention.

File: compiler/runtime.go [hunk 3]
```diff
@@ -520,7 +520,7 @@ func (e *identifierEvaluator) Eval(ctx context.Context, scope Scope) (values.Val
 	// value's **polytype** as this value will return an
 	// invalid monotype.
 	//
-	values.CheckKind(v.Type().Nature(), e.t.Nature())
+	// values.CheckKind(v.Type().Nature(), e.t.Nature())
```
[the remark this answers is not in the document]
Comment (jsternberg) [reply compiler/runtime.go]: Checking the kind just doesn't really work. I would have to add custom code for checking that the type was invalid or the nature and I think it's just easier to return the value as-is.

File: compiler/compiler.go (hunk 1 above)
Comment (jsternberg) [reply compiler/compiler.go]: Yes. This is to work around needing to mock out nulls everywhere in the binary lookup.

**Comment by jsternberg** [comment]
I thought making null an invalid type was easier than re-adding the null type which isn't and can't be used by Rust, which would mess up the semantic graph. It also has the benefit that if you have a null boolean and attempt to call `Bool()` it will panic which is correct behavior.

File: compiler/runtime.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jlapacik) [reply compiler/runtime.go]: Maybe we can just remove all of those then?

[the remark this answers is not in the document]
Comment (jsternberg) [reply compiler/runtime.go]: I think that's a good idea. We can do it as part of a clean up PR.

**Pull request merged**